src/main/java/io/oneko/kubernetes/impl/DeploymentManagerImpl.java

```diff
@@ -37,45 +45,85 @@ class DeploymentManagerImpl implements DeploymentManager {
 	private final DockerRegistryClientFactory dockerRegistryClientFactory;
 	private final ProjectRepository projectRepository;
 	private final DeploymentRepository deploymentRepository;
+	private final ProjectVersionLock projectVersionLock;
+	private final EventDispatcher eventDispatcher;
+
 
 	DeploymentManagerImpl(DockerRegistryClientFactory dockerRegistryClientFactory,
 												ProjectRepository projectRepository,
 												DeploymentRepository deploymentRepository,
-												EventDispatcher eventDispatcher) {
+												EventDispatcher eventDispatcher, ProjectVersionLock projectVersionLock) {
 		this.dockerRegistryClientFactory = dockerRegistryClientFactory;
 		this.projectRepository = projectRepository;
 		this.deploymentRepository = deploymentRepository;
+		this.projectVersionLock = projectVersionLock;
+		this.eventDispatcher = eventDispatcher;
 		eventDispatcher.registerListener(this::consumeDeletedVersionEvent);
 	}
 
 	@Override
-	public ReadableProjectVersion deploy(WritableProjectVersion version) {
+	public ReadableProjectVersion deploy(final WritableProjectVersion version) {
 		if (StringUtils.isBlank(version.getNamespaceOrElseFromProject())) {
 			throw new RuntimeException("A namespace must be configured in the project.");
 		}
-		try {
-			final UUID versionId = version.getId();
-			final WritableDeployment deployment = getOrCreateDeploymentForVersion(version);
 
-			if (!deployment.getReleaseNames().isEmpty()) {
-				HelmCommandUtils.uninstall(deployment.getReleaseNames());
-			}
+		final UUID versionId = version.getId();
 
-			final List<InstallStatus> installStatuses = HelmCommandUtils.install(version);
-			final List<String> releaseNames = installStatuses.stream().map(Status::getName).collect(Collectors.toList());
-			deployment.setReleaseNames(releaseNames);
-			deploymentRepository.save(deployment);
+		return projectVersionLock.doWithProjectVersionLock(version, () -> {
+			try {
+				final WritableDeployment deployment = getOrCreateDeploymentForVersion(version);
 
-			version = updateDeployableWithCreatedResources(version);
-			version.setDesiredState(Deployed);
+				if (!deployment.getReleaseNames().isEmpty()) {
+					HelmCommandUtils.uninstall(deployment.getReleaseNames());
+				}
 
-			final ReadableProject project = projectRepository.add(version.getProject());
-			return project.getVersions().stream()
-					.filter(projectVersion -> projectVersion.getUuid().equals(versionId))
-					.findFirst()
-					.orElse(null);
-		} catch (HelmRegistryException e) {
-			log.error("failed to deploy ({})", versionKv(version), e);
+				final List<InstallStatus> installStatuses = HelmCommandUtils.install(version);
+				log.info("Installing helm releases {} for {}",
```

Comment:
I'd prefer full lowercase log messages if there's just one sentence. What do you think?